design_drafts/non_joint_command_interfaces.md

```diff
@@ -0,0 +1,133 @@
+# Definition of command interfaces for non-joint related information
+
+## Motivation
+Already multiple times, users raised the question about the handling of non-joint-related interfaces.
+Such interfaces are not considerd during [design of the ros2_control URDF-tag structure](./components_architecture_and_urdf_examples.md).
+The design proposes two possibilities to classify (group) the interfaces, using `<joint>` and `<sensor>` tags.
+`<joint>`-tag groups the interfaces associated with the joints of physical robots and actuators.
+They have command and state interfaces to set the goal values for hardware and read its current state.
+`<sensor>`-tag groups multiple state interfaces describing, e.g., internal states of hardware.
+
+Most modern robots, especially industrial manipulators, also have other physical ports to control external hardware.
+The most usual ones are General Purpose Inputs/Outputs (GPIOs).
+Although the strict definition considers only digital ports, the term GPIO in this document describes any (one-dimensional) input/output port of physical hardware, e.g., digital or analog input/output.
+
+## Problem
+To use GPIOs in the ros2_control framework, they have to be listed in the robot's URDF: inputs (state interfaces) under a `<sensor>`-tag and output (command and state interfaces) under a `<joint>`-tag.
+This could lead to confusion since non-existing (virtual) joints have to be defined in the ros2_control part of the URDF structure. 
+"Joint" is a well-defined term in robotics. Therefore this naming should be avoided, i.e., another keyword should be used describing GPIOs,  semantically separating them from the kinematics joints.
+
+## Proposed solution
+Use of keyword "gpio" when describing input and output ports of a robotic device that cannot be associated with any joint or sensor.
+Parsing of `<gpio>-tag` is similar to this of a `<joint>`-tag having command and state interfaces.
+The tag must have at least one `<command>`- or `<state>`-tag as a child.
+
+The keyword "gpio" is chosen for its generality.
+Although strictly used for digital signals, in this document describes any electrical analog, digital signal, or physical value.
```

Comment:
GPIO as an acronym has a fairly well established meaning. See [wikipedia/GPIO](https://en.wikipedia.org/wiki/General-purpose_input/output) for instance.

Is the intention to use it like that in `ros2_control`, or is it even more *general*?

The text is slightly ambiguous, as in some places it states that the `gpio` tag should/could be used for *anything* which is not a `joint` or a `sensor`, but in other places (like here) it limits it to the more traditional meaning of the term (ie: "an uncommitted digital signal pin on an integrated circuit or electronic circuit board which may be used as an input or output, or both, and is controllable by the user at runtime" (Wikipedia)).

Reply:
@gavanderhoorn This is a very good point.

For clarification. We would like to use it for everything which has not "a physical" signal. The usual use-case would be digital and analog input outputs on a robot that can not be related to any joint or sensors.

> Is the intention to use it like that in ros2_control, or is it even more general?

The use is more general because it also describes analog signals.

Reply:
> We would like to use it for everything **which has not "a physical" signal**.

That seems to contradict what's in the document right now (emphasis mine):

> Although strictly used for digital signals, in this document describes any electrical analog, digital signal, **or physical value**.

> The usual use-case would be digital and analog input outputs on a robot that can not be related to any joint or sensors.

well, anything which you connect to a digital computer in the end is represented by *digital input outputs*. So this doesn't seem to say very much.

I understand why you went with `gpio`, but to me, it it's either not specific enough, or it's too generic.

And the given rationale and use-case seem a bit narrow as well: why only robots? What about using `ros2_control` for something which is not (directly) considered a robot?